interfaces/builtin/location.go

```diff
@@ -0,0 +1,241 @@
+  dbus (bind)
+      bus=system
+      name="com.ubuntu.location.Service",
+
+  dbus (bind)
+      bus=system
+      name="com.ubuntu.location.Service.Session",
+
+  # Allow traffic to/from our path and interface with any method
+  dbus (receive, send)
+      bus=system
+      path=/com/ubuntu/location/Service{,/**}
+      interface=com.ubuntu.location.Service*,
+
+  dbus (receive, send)
+      bus=system
+      path=/sessions/*
+      interface=com.ubuntu.location.Service.Session
+      member={Start,Stop}PositionUpdates,
+
+  dbus (receive, send)
+      bus=system
+      path=/sessions/*
+      interface=com.ubuntu.location.Service.Session
+      member={Start,Stop}HeadingUpdates,
+
+  dbus (receive, send)
+      bus=system
+      path=/sessions/*
+      interface=com.ubuntu.location.Service.Session
+      member={Start,Stop}VelocityUpdates,
+
+  # Allow traffic to/from org.freedesktop.DBus for location service
+  dbus (receive, send)
+      bus=system
+      path=/
+      interface=org.freedesktop.DBus**,
+  dbus (receive, send)
+      bus=system
+      path=/com/ubuntu/location/Service{,/**}
+      interface=org.freedesktop.DBus**,
+`)
+
+var locationConnectedPlugAppArmor = []byte(`
+# Description: Allow using location service. Reserved because this gives
+#  privileged access to the service.
+# Usage: reserved
+
+#include <abstractions/dbus-strict>
+
+# Allow all access to location service
+dbus (receive, send)
+    bus=system
+    peer=(label=###SLOT_SECURITY_TAGS###),
+
+dbus (send)
+    bus=system
+    peer=(name=com.ubuntu.location.Service, label=unconfined),
+
+dbus (send)
+    bus=system
+    peer=(name=com.ubuntu.location.Service.Session, label=unconfined),
+
+dbus (receive)
+    bus=system
+    path=/
+    interface=org.freedesktop.DBus.ObjectManager
+    peer=(label=unconfined),
+
+dbus (receive)
+    bus=system
+    path=/com/ubuntu/location/Service{,/**}
+    interface=org.freedesktop.DBus*
+    peer=(label=unconfined),
+`)
+
+var locationPermanentSlotSecComp = []byte(`
+getsockname
+recvmsg
+sendmsg
+sendto
+`)
+
+var locationConnectedPlugSecComp = []byte(`
+getsockname
+recvmsg
+sendmsg
+sendto
+`)
+
+var locationPermanentSlotDBus = []byte(`
+<policy user="root">
+	<allow own="com.ubuntu.location.Service"/>
+	<allow send_destination="com.ubuntu.location.Service"/>
+    <allow send_destination="com.ubuntu.location.Service.Session"/>
+	<allow send_interface="com.ubuntu.location.Service"/>
+	<allow send_interface="com.ubuntu.location.Service.Session"/>
+</policy>
+`)
+
+var locationConnectedPlugDBus = []byte(`
+<policy context="default">
+    <deny own="com.ubuntu.location.Service"/>               
+	<allow own="com.ubuntu.location.Service.Session"/>
+    <allow send_destination="com.ubuntu.location.Service"/>
+    <allow send_destination="com.ubuntu.location.Service.Session"/>
+    <allow send_interface="com.ubuntu.location.Service"/>
+	<allow send_interface="com.ubuntu.location.Service.Session"/>
```

Comment:
In addition to the whitespace issues, how this is supposed to work is that the server side binds to com.ubuntu.location.Service and the client side to com.ubuntu.location.Service.Session. Clients talk to com.ubuntu.location.Service and the service can talk to clients via com.ubuntu.location.Service.Session?

If that is accurate, why does PermanentSlotAppArmor allow the server to bind to both? Shouldn't it only allow binding to com.ubuntu.location.Service and the ConnectedPlugAppArmor policy allow binding to com.ubuntu.location.Service.Session? Otherwise, can you elaborate on how this works?